peg/elm.peg

```diff
@@ -0,0 +1,622 @@
+# Copyright (c) 2022 Nik Silver
+#
+# This source code is released for free distribution under the terms of the
+# GNU General Public License version 2 or later.
+#
+# This Elm parser will tag items reliably at the top level. Functions
+# defined in let/in blocks are also tagged, but with limitations. See below.
+#
+# Kinds
+# - m module
+# - n namespace (ie a module that's renamed)
+# - t type
+# - c constructor (within a type)
+# - a alias
+# - p port
+# - f function
+#
+# Key/value pairs
+# - roles:def       This is defined here.
+# - roles:imported  This is imported here.
+# - type:<t>        This constructor is in the scope of type <t>, which
+#                   may be dotted. Eg Main.myType.
+# - function:<f>    This function is in the scope of function <f>, which
+#                   may be dotted. Eg Main.myFunc.
+# - module:<m>      This is in the scope of module <m>.
+# - signature:<s>   This function, constructor or port has signature <s>.
+# - moduleName:<m>  This namespace has original module name <m>.
+#
+# Functions defined in let/in blocks may be tagged, with these limitations:
+# - the LHS (up to and including the '=') need to be on a single line;
+# - the LHS can only have simple parameters;
+# - their scope is only marked as being in the top-most function;
+# - any type annotation is ignored.
+# This should be good for 90% of inner functions. To make it totally robust
+# is much more complicated due to (a) Elm's clever indentation-sensitivity
+# and (b) limitations of the PEG parser used here.
+#
+# To do:
+# Maybe do:
+# - let/in blocks
+#   - Allow tuples on the LHS. Eg '(val1, val2) = valFunc'.
+#   - Inner functions' type annotations are used in the function's signature.
+#   - Inner functions can have more complex parameters.
+# - Functions
+#   - Allow non-Latin upper and lower case. Use
+#     https://util.unicode.org/UnicodeJsps/properties.html
+#     combined with \p{Lu}, \p{Ll} and \p{L}.
+# - If there's no --field=+r then an imported module (say) is still tagged,
+#   but there's no roles:imported, so it looks like it has been defined.
+#   Not sure if this is a bug. The Thrift parser has this same behaviour.
+#
+# Won't do:
+# - Handle Elm's indentation properly.
+
```

Comment:
Could you put the URL for a reference of the target language as Jan did in kotlin.peg?